README.md
# connect-more

Variation of the popular two-player board game Connect Four, implemented in Haskell using Gloss.

![Sample image](sample.png)

## Building/Running

With GHC, cabal and stack installed, the library can be built by running `stack build` in a Terminal window.

The game can then be run locally using the command: `stack run`.


## Gameplay 

The gameplay follows that of the original game, wherein players take turns adding a circular piece to an unfilled column in the board, until one player wins by getting the chosen number of pieces of the same colour in a row (vertically, horizontally or diagonally); or until the board is filled, in which case a draw is declared.

The main extension to the original game is the ability to choose the **dimensions of the board** (each side limited to between 3 and 9 pieces) as well as the **number of pieces in a row** required to win (limited to between 3 and the minimum out of the number of rows and columns), via the settings menu&mdash;hence the name of my version of the game.

Another desirable feature is the alternation of the starting players; if Red starts one game, Blue will start the next, in accordance with the original game.

## Controls (keyboard)

- s: settings
- h: show/hide help menu
- q: quit to main menu, reset score
- numbers 1-no. of columns: add piece to column
- left arrow key: undo previous move
- spacebar: start new game (with default settings if on main menu)



ChangeLog.md
# Changelog for connect-more

## Unreleased changes


app/Main.hs
{-# LANGUAGE RecordWildCards #-}

module Main where

import Graphics.Gloss hiding (Point)
import Graphics.Gloss.Interface.IO.Game hiding (Point)
import Data.Char (intToDigit, digitToInt)
import Data.Map (Map, (!))
import qualified Data.Map as Map

import Config
import Board
import Renders

-- Data type defining different game states/screens
data State = Title | SetCols | SetRows | SetWin | Help | Play | Endgame
    deriving (Eq)

-- Data type storing game information
data World = World {
    state     :: State,          -- Current state of game
    prevState :: State,          -- Previous state (recorded to return from help menu)
    win       :: Int,            -- Number of pieces in a row to win
    board     :: Board,          -- Current board
    moves     :: [Point],        -- Record of moves made in order
    player    :: Piece,          -- Current player
    starter   :: Piece,          -- Starter of next game (alternates every game)
    winner    :: Piece,          -- Winning player
    score     :: Map Piece Int   -- Record of scores (players are keys, games won are values)
}

-- Initial World structure rendered when game starts
initialWorld :: World
initialWorld = World {
    state     = Title,           -- Start on the title screen
    prevState = Title,           -- Set previous state to title
    win       = 4,               -- Default number of pieces in a row to win
    board     = emptyBoard 7 6,  -- Reset board with default columns and rows
    moves     = [],              -- No moves made
    player    = Red,             -- Red starts first game
    starter   = Blue,            -- Blue starts next game
    winner    = None,            -- No winner yet
    score     = initialScore     -- All scores set to 0
}


-- Main entry point to program, handles gameplay
main :: IO ()
main = play
    (InWindow "Connect More" windowSize windowPos) -- Display mode
    white                                          -- Background colour
    60                                             -- Updates per second of real time
    initialWorld                                   -- Initial world
    render                                         -- Rendering function
    handleInput                                    -- Input handler
    update                                         -- Update handler


-- Render world to picture based on state of game
render :: World -> Picture
render World{..} = 
    let 
        -- Maximum number of pieces in a row to win limited by smallest board dimension
        maxWin = min (cols board) (rows board)
    in 
        pictures $ case state of
        Title   -> renderTitle
        SetCols -> renderSettings board win maxWin 152
        SetRows -> renderSettings board win maxWin 92
        SetWin  -> renderSettings board win maxWin 32
        Help    -> renderHelp
        Play    -> renderPlay score win board player
        Endgame -> renderEndgame score win board winner


-- Updates world based on user's interactions with game through key presses
handleInput :: Event -> World -> World

-- Handle spacebar press - start new game
handleInput (EventKey (SpecialKey KeySpace) Down _ _) world@World{..} = 
    case state of
        Title   -> world { state   = Play }                  -- Start new game
        Endgame -> world { state   = Play,                   -- Start new game
                           board   = emptyBoard (cols board) (rows board), -- Reset board, maintaining size
                           moves   = [],                     -- Reset move history
                           player  = starter,                -- Alternate starting player
                           starter = opposite starter,       -- Opposite player starts next game
                           winner  = None                    -- Reset winner
                         }
        Help    -> world                                     -- Ignore if on help menu
        Play    -> world                                     -- Ignore if playing game
        _       -> world { state = Play }                    -- Start game from settings menu

-- Handle down arrow key press - navigate settings menu
handleInput (EventKey (SpecialKey KeyDown) Down _ _) world@World{..} = 
    case state of
        SetCols  -> world { state = SetRows }  -- Move down settings menu to set number of rows
        SetRows  -> world { state = SetWin }   -- Move down settings menu to set number of pieces in a row to win
        _        -> world                      -- Ignore if not on settings screen

-- Handle up arrow key press - navigate settings menu
handleInput (EventKey (SpecialKey KeyUp) Down _ _) world@World{..} =
    case state of
        SetRows -> world { state = SetCols }   -- Move up to settings menu to number of columns
        SetWin  -> world { state = SetRows }   -- Move up to settings menu to number of rows
        _       -> world                       -- Ignore if not on settings screen

-- Handle left arrow key press - undo previous move
handleInput (EventKey (SpecialKey KeyLeft) Down _ _) world@World{..} =
    case state of
        Play | not $ null moves -> world { board  = undoMove board (last moves),      -- Remove last piece added until empty
                                           moves  = init moves,                       -- Remove record of last move
                                           player = opposite player                   -- Go back to previous player
                                         }
        Endgame                 -> world { state  = Play,                             -- Resume previous game
                                           board  = undoMove board (last moves),      -- Remove last piece added
                                           moves  = init moves,                       -- Remove record of last move
                                           winner = None,                             -- Undeclare winner
                                           score  = Map.adjust (+ (-1)) winner score  -- Decrement winner's score
                                         }
        _                       -> world                                              -- Ignore in other states

-- Handle character key presses
handleInput (EventKey (Char c) Down _ _) world@World{..} = 
    let cs = cols board -- Shorthand for number of columns
        rs = rows board -- Shorthand for number of rows
    in case state of
        Title   | c == 's'                                 -> world { state = SetCols }                     -- Go to settings menu
        Play    | c `elem` ['1'..(intToDigit cs)]          -> makeMove world (digitToInt c)                 -- Add piece to column entered
        SetCols | c `elem` ['3'..'9']                      -> world { state = SetRows,                      -- Move down to next setting
                                                                      board = emptyBoard (digitToInt c) rs, -- Set new number of columns
                                                                      win   = win' win (digitToInt c) rs    -- Update pieces to win if necessary
                                                                    }
        SetRows | c `elem` ['3'..'9']                      -> world { state = SetWin,                       -- Move down to next setting
                                                                      board = emptyBoard cs (digitToInt c), -- Set new number of rows
                                                                      win   = win' win cs (digitToInt c)    -- Update pieces to win if necessary
                                                                    }
        SetWin  | c `elem` ['3'..(intToDigit $ min cs rs)] -> world { win   = digitToInt c }                -- Set new number of rows
        _       | c == 'h', state /= Help                  -> world { state = Help, prevState = state }     -- Show help menu
        _       | c == 'h'                                 -> world { state = prevState }                   -- Return to previous screen from help menu
        _       | c == 'q'                                 -> initialWorld                                  -- Quit to title screen
        _                                                  -> world                                         -- Ignore other character key presses

-- Ignore any other inputs
handleInput _ world = world

-- Attempt to add a piece to the column entered
makeMove :: World -> Int -> World
makeMove world@World{..} col =
    let
        row = heights board ! col + 1                 -- Get row to add piece to (column height + 1)
        newBoard = updateBoard board (col,row) player -- Add piece to board in selected column
    in
        -- Update game if column to add piece to is not full, otherwise do not amke move
        if row <= rows board then
            update 0 world { board = newBoard, moves = moves ++ [(col,row)] }
        else world


-- Update world when player makes a move
update :: Float -> World -> World
update 0 world@World{..}

    -- Declare draw if board is full (no unfilled columns)
    | boardIsFull board       = world { state  = Endgame,
                                        score  = Map.adjust (+ 1) None score }
    -- End game if player has won
    | hasWon board win player = world { state  = Endgame,
                                        winner = player,
                                        score  = Map.adjust (+ 1) player score }
    -- Continue playing game, set next player
    | otherwise               = world { state  = Play,
                                        player = opposite player }

-- Do nothing unless called after making a move
update _ world = world


-- Return the board after removing the piece at a given position
undoMove :: Board -> Point -> Board
undoMove board@Board{..} coords = board {
    pieces  = Map.insert coords None pieces,            -- Remove piece at position (add None)
    heights = Map.adjust (+ (-1)) (fst coords) heights  -- Decrement height of column
    }

-- Limit the maximum number of pieces in a row to win to the smallest board dimension
win' :: Int -> Int -> Int -> Int
win' currWin cols rows = if currWin > maxWin then maxWin else currWin
    where maxWin = min cols rows

src/Board.hs
{-# LANGUAGE RecordWildCards #-}

module Board where

import qualified Data.Map as Map
import Data.Map (Map, (!))
import Data.List (isSubsequenceOf)

-- Position in a board represented as a (column, row) pair
type Point = (Int,Int)

-- Represention of the board to which pieces are added
data Board = Board { pieces  :: Map Point Piece, -- Map storing the piece (value) at each position (key)
                     heights :: Map Int Int,     -- Map storing the height/number of pieces (value) in each column (key)
                     cols    :: Int,             -- Number of columns
                     rows    :: Int }            -- Number of grids

-- Representation of pieces to be added to board, also used to represent players
data Piece = None | Red | Blue
    deriving (Show, Eq, Ord)

-- Empty board with no pieces placed
emptyBoard :: Int -> Int -> Board
emptyBoard cols rows = Board {
    pieces  = Map.fromList [((x,y), None) | x <- [1..cols], y <- [1..rows]],
    heights = Map.fromList [(col,0) | col <- [1..cols]],
    cols = cols,
    rows = rows
    }

-- Initial score - all individual scores set to 0
initialScore :: Map Piece Int
initialScore = Map.fromList [(piece, 0) | piece <- [None, Red, Blue]]

-- Get the piece at a given position in the board
getPiece :: Board -> Point -> Piece
getPiece board coords = pieces board ! coords

-- Get the opposite piece/player 
opposite :: Piece -> Piece
opposite Red  = Blue
opposite Blue = Red
opposite None = None

-- Update the board after adding a piece at a given position
updateBoard :: Board -> (Int,Int) -> Piece -> Board
updateBoard Board{..} (col,row) player = Board {
    pieces  = Map.insert (col,row) player pieces,
    heights = Map.insert col row heights,
    cols = cols,
    rows = rows
    }

-- Return if the board is full by checking if there are no unfilled columns
boardIsFull :: Board -> Bool
boardIsFull Board{..} = null $ Map.filter (< rows) heights

-- Return if a player has won by checking if their occupied positions include any winning combinations
hasWon :: Board -> Int -> Piece -> Bool
hasWon board@Board{..} win player = any (`isSubsequenceOf` occupied board player) allWins
    where
        -- All winning combinations based on size of board and number of pieces in a row to win
        allWins =
            let
                -- Get possible winning combinations for each direction
                horizontalWins = [ [(c + z, r)     | z <- [0..(win - 1)]] | c <- [1..(cols - (win - 1))], r <- [1..rows] ]
                verticalWins   = [ [(c, r + z)     | z <- [0..(win - 1)]] | c <- [1..cols], r <- [1..(rows - (win - 1))] ]
                leftDiagWins   = [ [(c + z, r - z) | z <- [0..(win - 1)]] | c <- [1..(cols - (win - 1))], r <- [win..rows] ]
                rightDiagWins  = [ [(c + z, r + z) | z <- [0..(win - 1)]] | c <- [1..(cols - (win - 1))], r <- [1..(rows - (win - 1))] ]
            in
                -- Combine winning combinations for each direction into single list
                verticalWins ++ horizontalWins ++ leftDiagWins ++ rightDiagWins

-- Get list of positions occupied in the board by a given player
occupied :: Board -> Piece -> [Point]
occupied board player = Map.keys $ Map.filter (== player) (pieces board)

src/Config.hs
module Config where

import Graphics.Gloss

import Board ( Piece(..) )

-- Size of game window in pixels
windowSize :: (Int,Int)
windowSize = (900,830)

-- Size of board in pixels
boardSize :: (Int,Int)
boardSize = (650,550)

-- Accessors for board dimensions in pixels
boardWidth, boardHeight :: Int
boardWidth  = fst boardSize
boardHeight = snd boardSize

-- Position window in centre of monitor (on 14-inch Macbook)
windowPos :: (Int,Int)
windowPos  = (310,200)

-- Scale piece radius appropriately as a function of the number of columns and rows
pieceRadius :: Int -> Int -> Float
pieceRadius cols rows = 210 / fromIntegral (max cols rows)

-- Calculate pixels occupied by each piece
xUnit, yUnit :: Int -> Float
xUnit cols = fromIntegral boardWidth  / fromIntegral cols
yUnit rows = fromIntegral boardHeight / fromIntegral rows

-- Grey colour used for text
grey :: Color
grey = greyN 0.4

-- Get the colour of each type of piece
getColour :: Piece -> Color
getColour None = makeColor 0.1 0.2 0.3 0.1
getColour Red  = makeColor 1 0.4 0.4 0.8
getColour Blue = makeColor 0.4 0.6 1 0.9

-- Get the background colour to display when a player wins
getBG :: Piece -> Color
getBG None = greyN 0.9
getBG Red  = makeColor 1 0.4 0.4 0.3
getBG Blue = makeColor 0.4 0.6 1 0.3

src/Renders.hs
{-# LANGUAGE RecordWildCards #-}

module Renders where

import Graphics.Gloss
import Data.Map (Map, (!))

import Config
import Board

-- Helper function for rendering text
-- Source: https://github.com/dixonary/hake
text' :: Color -> Float -> Float -> String -> Picture
text' col size boldness str =
    color col $
    pictures [translate x y $ scale (size / 100) (size / 100) $ text str
         | x <- [(-boldness),(0.5-boldness)..boldness],
           y <- [(-boldness),(0.5-boldness)..boldness] ]

-- x-coordinate for left-aligning text in setup/help menus
alignL :: Float
alignL = -360

-- Render title screen with logo and text
renderTitle :: [Picture]
renderTitle = [
    translate (-35) 100 $ color (getColour Red) $ circleSolid 50,
    translate 35 100  $ color (getColour Blue) $ circleSolid 50,
    translate (-175) (-20)  $ text' grey 40 1   "connect more",
    translate (-223) (-80) $ text' grey 25 0.8 "Press space to start game",
    translate (-160) (-130) $ text' grey 20 0.8 "Press s to view settings",
    translate (-140) (-180) $ text' grey 20 0.8 "Press h to view help"
    ]

-- Render settings menu with board dimensions, pieces in a row to win and instructions
renderSettings :: Board -> Int -> Int -> Float -> [Picture]
renderSettings Board{..} win maxWin rectY = [
    translate 0 rectY $ color (greyN 0.95) $ rectangleSolid 800 60, -- Rectangle indicating selected field
    translate alignL 290 $ text' grey 40 1.2 "Game settings",
    translate alignL 210 $ text' grey 25 1 "Enter number within range to change",
    translate alignL 140 $ text' grey 25 1 ("- Number of columns: " ++ show cols ++ "    (3-9)"),
    translate alignL 80  $ text' grey 25 1 ("- Number of rows: " ++ show rows ++ "      (3-9)"),
    translate alignL 20  $ text' grey 25 1 ("- Pieces in a row to win: " ++ show win  ++ " (3-" ++ show maxWin ++ ")"),
    translate alignL (-100) $ text' grey 25 1 "Navigate with up/down arrow keys",
    translate alignL (-160) $ text' grey 25 1 "Press space to start game"
    ]

-- Render help menu with gameplay instructions and controls
renderHelp :: [Picture]
renderHelp = [
    translate alignL 290    $ text' grey 40 1.2 "Help menu",
    translate alignL 210    $ text' grey 25 1.2 "Gameplay",
    translate alignL 150    $ text' grey 20 0.8 "- Players take turns adding one piece to an unfilled",
    translate alignL 100    $ text' grey 20 0.8 "  column, until one player wins by placing x pieces",
    translate alignL 50     $ text' grey 20 0.8 "  in a row (x specified by user during setup).",
    translate alignL 0      $ text' grey 20 0.8 "- Draw is declared when the board is filled.",
    translate alignL (-70)  $ text' grey 25 1.2 "Controls (keyboard)",
    translate alignL (-130) $ text' grey 20 0.8 "- Numbers 1-7: add pieces to column",
    translate alignL (-180) $ text' grey 20 0.8 "- Left arrow key: undo previous move",
    translate alignL (-230) $ text' grey 20 0.8 "- q: quit to main menu, reset score",
    translate alignL (-280) $ text' grey 20 0.8 "- h: show/hide help menu"
    ]

-- Render ongoing game with score, pieces to win, board, column numbers and current player
renderPlay :: Map Piece Int -> Int -> Board -> Piece -> [Picture]
renderPlay score win board player = [
    renderScore score,
    renderGoal win,
    renderBoard board,
    renderColNums board,
    translate (-150) (-370) $ text' grey 22 0.8 "Current player: ",
    translate 72 (-370) $ text' (getColour player) 22 0.8 $ show player
    ]

-- Render current score for each player, including number of draws
renderScore :: Map Piece Int -> Picture
renderScore score = translate (-184) 350 $ text' grey 20 0.8 (
    "Draw: " ++ show (score ! None) ++ " | " ++
    "Red: "  ++ show (score ! Red)  ++ " | " ++
    "Blue: " ++ show (score ! Blue)
    )

-- Render current numbers of pieces in a row to win, with help instructions
renderGoal :: Int -> Picture
renderGoal win = translate (-218) 300 $ text' grey 22 0.8 (
    "Connect " ++ show win ++ " (press h for help)")

-- Render board in the centre of the game window
-- Adapted from: https://github.com/dixonary/hake
renderBoard :: Board -> Picture
renderBoard board@Board{..} = pictures $ [renderPiece col row | col <- [1..cols], row <- [1..rows]]
    where
        renderPiece col row =
            let
                -- Midpoint of piece relative to board dimensions
                xRel = fromIntegral (col - 1) - (fromIntegral cols - 1) / 2
                yRel = fromIntegral (row - 1) - (fromIntegral rows - 1) / 2

                -- Position based on distance from middle
                xPos = xRel * xUnit cols
                yPos = yRel * yUnit rows
            in
                -- Display piece as circle in correct colour
                translate xPos yPos $ color (getColour piece) $ circleSolid (pieceRadius cols rows)
                    where
                         -- Get the piece at the current position
                        piece = getPiece board (col,row)

-- Render column numbers below each column for player's reference
renderColNums :: Board -> Picture
renderColNums Board{..} = pictures $ [renderNumber col | col <- [1..cols]]
    where
        renderNumber col =
            let
                xRel = fromIntegral (col - 1) - (fromIntegral cols - 1) / 2 -- Midpoint of number relative to number of columns
                xPos = xRel * xUnit cols - 11                               -- Position based on distance from middle
            in 
                translate xPos (-320) $ text' grey 30 1 (show col)          -- Display column number

-- Render endgame screen with score, board, winning player and instructions
renderEndgame :: Map Piece Int -> Int -> Board -> Piece -> [Picture]
renderEndgame score goal board winner = [
    translate 0 0 $ color (getBG winner) $ rectangleSolid 1000 1000, -- Change background colour depending on winner
    renderGoal goal,
    renderScore score,
    renderBoard board,
    renderWinner winner,
    translate (-163) (-370) $ text' grey 22 0.8 "Press space to restart"
    ]

-- Render text indicating winning player or draw
renderWinner :: Piece -> Picture
renderWinner None   = translate (-50) (-320) $ text' grey 30 1 "Draw!"
renderWinner player = translate (-90) (-320) $ text' (getColour player) 30 1 (show player ++ " wins!")